homeassistant/components/switchbot/cover.py

```diff
@@ -0,0 +1,139 @@
+"""Support for SwitchBot curtains."""
+from __future__ import annotations
+
+import logging
+from typing import Any
+
+from homeassistant.components.cover import (
+    ATTR_CURRENT_POSITION,
+    ATTR_POSITION,
+    DEVICE_CLASS_CURTAIN,
+    SUPPORT_CLOSE,
+    SUPPORT_OPEN,
+    SUPPORT_SET_POSITION,
+    SUPPORT_STOP,
+    CoverEntity,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_MAC, CONF_NAME, CONF_PASSWORD
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.restore_state import RestoreEntity
+
+from .const import CONF_RETRY_COUNT, DATA_COORDINATOR, DOMAIN
+from .coordinator import SwitchbotDataUpdateCoordinator
+from .entity import SwitchbotEntity
+
+# Initialize the logger
+_LOGGER = logging.getLogger(__name__)
+PARALLEL_UPDATES = 1
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up Switchbot curtain based on a config entry."""
+    coordinator: SwitchbotDataUpdateCoordinator = hass.data[DOMAIN][entry.entry_id][
+        DATA_COORDINATOR
+    ]
+
+    async_add_entities(
+        [
+            SwitchBotCurtain(
+                coordinator,
+                entry.unique_id,
+                entry.data[CONF_MAC],
+                entry.data[CONF_NAME],
+                entry.data.get(CONF_PASSWORD),
+                entry.options[CONF_RETRY_COUNT],
+            )
+        ]
+    )
+
+
+class SwitchBotCurtain(SwitchbotEntity, CoverEntity, RestoreEntity):
+    """Representation of a Switchbot."""
+
+    coordinator: SwitchbotDataUpdateCoordinator
+    _attr_device_class = DEVICE_CLASS_CURTAIN
+    _attr_supported_features = (
+        SUPPORT_OPEN | SUPPORT_CLOSE | SUPPORT_STOP | SUPPORT_SET_POSITION
+    )
+    _attr_assumed_state = True
+
+    def __init__(
+        self,
+        coordinator: SwitchbotDataUpdateCoordinator,
+        idx: str | None,
+        mac: str,
+        name: str,
+        password: str | None,
+        retry_count: int,
+    ) -> None:
+        """Initialize the Switchbot."""
+        super().__init__(coordinator, idx, mac, name)
+        self._attr_unique_id = idx
+        self._device = self.coordinator.switchbot_api.SwitchbotCurtain(
+            mac=mac, password=password, retry_count=retry_count
+        )
```

Comment:
It would be cleaner to create the device in `async_setup_entry` and pass it to `SwitchBotCurtainEntity`

Reply:
Done